src/senaite/core/exportimport/instruments/importer.py

```diff
@@ -0,0 +1,1049 @@
+                            mapping={"analysis_keyword": keyword}))
+                continue
+            # remember the valid service keyword
+            keywords.append(keyword)
+
+        if len(keywords) == 0:
+            self.warn(_("No services could be found for parsed keywords"))
+
+        return keywords
+
+    def is_valid_keyword(self, keyword):
+        """Check if the keyword is valid
+        """
+        results = self.setup_catalog(getKeyword=keyword)
+        if not results:
+            return False
+        return True
+
+    def get_reference_sample_by_id(self, sid):
+        """Get a reference sample by ID
+        """
+        query = {"portal_type": "ReferenceSample", "getId": sid}
+        results = api.search(query, SENAITE_CATALOG)
+        if len(results) == 0:
+            return None
+        return api.get_object(results[0])
+
+    def get_sample_by_id(self, sid):
+        """Get a  sample by ID
+        """
+        query = {"portal_type": "AnalysisRequest", "getId": sid}
+        results = api.search(query, SAMPLE_CATALOG)
+        if len(results) == 0:
+            return None
+        return api.get_object(results[0])
+
+    def get_attachment_type_by_title(self, title):
+        """Get an attachment type by title
+
+        :param title: Attachment type title
+        :returns: Attachment object or None
+        """
+        query = {
+            "portal_type": "AttachmentType",
+            "title": title,
+            "is_active": True,
+        }
+        results = self.setup_catalog(query)
+        if not len(results) > 0:
+            return None
+        return api.get_object(results[0])
+
+    def process(self):
+        parsed_results = self.parse_results()
+
+        # no parsed results, return
+        if not parsed_results:
+            return False
+
+        # Log allowed sample and analyses states
+        self.log(_("Allowed sample states: ${allowed_states}", mapping={
+            "allowed_states": ", ".join(self.allowed_sample_states_msg)
+        }))
+        self.log(_("Allowed analysis states: ${allowed_states}", mapping={
+            "allowed_states": ", ".join(self.allowed_analysis_states_msg)
+        }))
+        if not any([self.override_non_empty, self.override_with_empty]):
+            self.log(_("Don't override analysis results"))
+        if self.override_non_empty:
+            self.log(_("Override non-empty analysis results"))
+        if self.override_with_empty:
+            self.log(_("Override non-empty analysis results, also with empty"))
+
+        # Attachments will be created in any worksheet that contains
+        # analyses that are updated by this import
+        attachments = {}
+        infile = self.parser.getInputFile()
+
+        ancount = 0
+        updated_analyses = []
+        importedinsts = {}
+        importedars = {}
+
+        for sid, results in parsed_results.items():
+            # get the sample
+            sample = self.get_sample_by_id(sid)
```

Comment:
This line (and the function `get_sample_by_id` as well) can be removed cause the `sample` variable is never used in this function. System relies on `get_analyses_for` function to directly get the analyses, without the need of looking for the sample first.

Reply:
Good catch! Removed here: e37297118